### PR TITLE
Auto-detect MySQL flavor.

### PR DESCRIPTION
**Pull request by alainjobart**
The main change is that we get the flavor from an opened client connection to the server, instead of knowing it beforehand (or even guessing it). Then the flavor interface and code is in go/mysql now. This seems a lot cleaner and overall less lines of code.

### CODE REVIEW

**Comment by acharis** [comment]
sounds tasty

**Comment by alainjobart** [comment]
@sougou I'll trade you a review on this for a review on your PR. :)

Each commit is probably easier to review by itself, as each moves a functionality.

File: go/mysql/flavor.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+// SendBinlogDumpCommand sends the flavor-specific version of
+// the COM_BINLOG_DUMP command to start dumping raw binlog
+// events over a slave connection, starting at a given GTID.
+func (c *Conn) SendBinlogDumpCommand(slaveID uint32, startPos Position) error {
+	return c.flavor.sendBinlogDumpCommand(c, slaveID, startPos)
+}
+
+// ResetReplicationCommands returns the commands to completely reset
+// replication on the host.
+func (c *Conn) ResetReplicationCommands() []string {
```
Comment (sougou) [inline go/mysql/flavor.go]: One worry about this approach is that the Conn API may get crowded due to these flavor-specific abstractions. But this is still better than what we had before.

In the long run, we can consider supporting these through vitess-specific SQL statements (like `RESET REPLICATION`) that get translated into flavor-specific commands under the covers.

File: go/vt/dbconnpool/pooled_connection.go [hunk 2]
```diff
@@ -38,39 +31,14 @@ func (pc *PooledDBConnection) Recycle() {
...
-// of DBConnection objects.
-//
-// For instance:
-// mysqlStats := stats.NewTimings("Mysql")
-// pool := dbconnpool.NewConnectionPool("name", 10, 30*time.Second)
-// pool.Open(dbconnpool.DBConnectionCreator(info, mysqlStats))
-// ...
-// conn, err := pool.Get()
-// ...
-func DBConnectionCreator(info *mysql.ConnParams, mysqlStats *stats.Timings) CreateConnectionFunc {
```
Comment (sougou) [inline go/vt/dbconnpool/pooled_connection.go]: good riddance!

**Pull request merged**